### ISSUE TITLE
Control homebridge with the command line (without an iPhone)

### ISSUE THREAD

**Issue by oa-** [nfarina/homebridge#506]
When setting up a new configuration I'd like to test it without using Siri all the time.

How can I simulate the command Siri is sending to homebridge simply using the terminal? Thanks.

**Comment by KhaosT** [comment]
Why don't you just test the homebridge with a HomeKit app...?

**Comment by oa-** [comment]
…oh, right.

In fact I want to send commands to homebridge **without an iPhone**/iOS device.

Is there a way to do so from a Mac running OS X?

**Comment by cflurin** [comment]
@oa: Take a look [here](https://www.npmjs.com/package/homebridge-punt).
homebridge-punt has a build-in simulator. The simulator works in both directions ios-device <-> gateway.

**Comment by oa-** [comment]
Thank you, I am currently testing homebridge-punt. As of yet I wasn't able to see my other devices in the punt simulator.

Other hints to control homebridge are welcome as well.
(e.g. something like `homebridge --setAccessory "Light 1" --setCharacteristic "on"`)

**Comment by yene** [comment]
You can run the apple example app in Simulator. https://developer.apple.com/library/ios/samplecode/HomeKitCatalog/Introduction/Intro.html

**Comment by KhaosT** [comment]
@oa- or you can use curl and simulate HomeKit request. For this option, you may want to check [here](https://homebridgeteam.slack.com/archives/random/s1453833581000293)

**Comment by cflurin** [comment]
@oa- that's right, the punt simulator is only for the accessories defined in config-punt.json but that's good point. However I'm not sure whether I can access all the homebridge accessories in my plugin. I've to investigate.

**Comment by rooi** [comment]
@KhaosT the simulation of a homekit request sounds interesting. However, the link you included is not accessible without a login. Do you have another link or know how I can get it another way? Thanks!

**Comment by KhaosT** [comment]
@rooi You can join the homebridge channel from the README.
Here are the contents I'm referring:
```
PUT http://<homebridge ip>:51826/characteristics` accepts JSON with a little array of values you wish to set
so to turn a light on: `{ characteristics: [ { aid: 1, iid: 8, value: true } ] }`
you can look through `/accessories` JSON to figure out what the `aid` and `iid` values should be
```
And if you're using homebridge 0.2.16, you may need to manually edit the hap-nodejs and comment out this part (https://github.com/KhaosT/HAP-NodeJS/blob/3733e7d7274a656c71fe02fd3df4af170dd33c3b/lib/HAPServer.js#L700)

**Comment by desvdp** [comment]
Hi @KhaosT  I tried to look at the homebridgeteam.slack but it seems you need to be invited ... could you invite me? Thx !

**Comment by KhaosT** [comment]
Check the project's GitHub page, there is a slack button. Click that and you'll get invitation to the channel ^^

**Comment by desvdp** [comment]
@KhaosT Thanks for the hint ! I missed the link the first time...

**Comment by rooi** [comment]
@KhaosT This has changed in the latest HAP-NodeJS version right? If I understand the code correctly you should be able to set the auth en get a response. If this is correct, could you please indicate what the commands are?

I get Insufficient privileges with something like (with different pincode "\"' variations and other options):

curl -X PUT http://mac-mini.local:51828/accessories --header "Content-Type:Application/json" --header "authentication: \"031-45-154\"" -s

Thanks!

**Comment by KhaosT** [comment]
Make sure you're on latest homebridge (0.2.18). Launch homebridge with flag ``-I`` and replace "authentication" with "authorization".

**Comment by rooi** [comment]
It works, thanks!

**Comment by derekmoyes** [comment]
Also, you could try this:
https://github.com/KraigM/HomeBridgeController 
works for me, ymmv. :-)

**Comment by Samfox2** [comment]
@KhaosT I'd like to set a value of an accessory within a bash script. I got the aai/iid for my accessory as discribed above. However, I'm struggling with the correct syntax of the curl command for PUTing a value to the accessory. Could you please indicate how to do this?

My (not working) command is:
`curl -X PUT http://localhost:PORT/characteristics --header "Content-Type:Application/json" --header "authorization: \"031-45-XXX\"" --data "{\"aid\":\"AID\",\"iid\":\"IID\",\"value\":\"VALUE\"}"`

**Comment by snowdd1** [comment]
@Samfox2 You missed the "characteristics" part that surrounds your data, see khaosT's post
https://github.com/nfarina/homebridge/issues/506#issuecomment-180849887 from Feb 6

**Comment by Samfox2** [comment]
Added it now but still no change:

`curl -X PUT http://localhost:PORT/characteristics --header "Content-Type:Application/json" --header "authorization: \"031-45-XXX\"" --data "{ characteristics: [{ \"aid\": \AID\", \"iid\": \"IIDl\", \"value\": \"VALUE\"}] }"
`

**Comment by KhaosT** [comment]
@Samfox2 you need to post what's the error you get...

**Comment by Samfox2** [comment]
The error is:

`{"status":-70401}`

**Comment by KhaosT** [comment]
@Samfox2 that means you are not running homebridge in "insecure" mode. In order to send command via a non-HomeKit path, you need to run homebridge with [``-I`` option](https://github.com/nfarina/homebridge/blob/master/lib/cli.js#L20).

**Comment by Samfox2** [comment]
Hmm, strange, I thought I did as it was working with the accessories-method

**Comment by KhaosT** [comment]
@Samfox2 Huh interesting. If you are sure it's running with ``-I`` option then the pincode you send is incorrect... [Here](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/HAPServer.js#L948) is how I check that.

**Comment by Samfox2** [comment]
No, it's not working 😔
- started the session with -I 
- hardcoded  this.allowInsecureRequest = true;
- commented out the section with autentification

Would you mind checking it with your config and posting the whole command? Maybe there is a syntax error in my command?

**Comment by KhaosT** [comment]
It looks right to me... Try add stuff to print out the header and see if it actually getting the correct header?

``curl -X PUT http://ti.io/characteristics --header "Content-Type:application/json" --header "Authorization: \"031-45-154\""``

**Comment by Samfox2** [comment]
Looks like there are a few spelling errors in HAPServer.js.
If I change all "authorization" strings to "Authorization" the error message disappears (although it is not working yet).

**Comment by Samfox2** [comment]
After a long debug session the following command deliveres the correct header incl. pin but characteristics is empty {}:

`curl -X PUT http://localhost:11111/characteristics --header "Content-Type:Application/json" --header "authorization: 031-11-111" --data "aid: 111" --data "iid: 1" --data "value: 1" -s`

Any idea?

**Comment by KhaosT** [comment]
@Samfox2 it expects a json payload ... not ``--data "aid: 111" --data "iid: 1" --data "value: 1"`` ...

**Comment by Samfox2** [comment]
Sorry, its too late...

I mean:


`curl -X PUT http://localhost:11111/characteristics --header "Content-Type:Application/json" --header "authorization: 031-11-111" --data "{ characteristics: [{ \"aid\": \"111\", \"iid\": \"1\", \"value\": \"1\"}] }"`

**Comment by KhaosT** [comment]
Try ``curl -X PUT http://localhost:11111/characteristics --header "Content-Type:Application/json" --header "authorization: 031-11-111" --data "{ \"characteristics\": [{ \"aid\": \"111\", \"iid\": \"1\", \"value\": \"1\"}] }"`` ?

**Comment by Samfox2** [comment]
Got it to work, finally...
My physical doorbell triggeres now with a script my homekit doorbell and sends a notification + pic right to my ios devices 😁

The correct command is:

`curl -X PUT http://localhost:PORT/characteristics --header "Content-Type:Application/json" --header "authorization:031-45-111" --data "{\"characteristics\":[{\"aid\":AID,\"iid\":IID,\"value\":VALUE}]}"
`

Thanks for your help!

**Comment by gijoecool** [comment]
Is the pic part of the notification?

**Comment by Samfox2** [comment]
Yes, pic and videostream if the pic is clicked.

**Comment by cpoetter** [comment]
Hi @Samfox2 
May I ask how you got the AID and IID? I tried what @KhaosT wrote (commented lines out) and also looked into the accessories directory, but I can't find any clue of AID and IID...
Thanks for your help!

**Comment by Samfox2** [comment]
The following bash script should save all info to a file. To improve readability I used jp but you can skip it if you do not want to use it.
[https://github.com/paulhammond/jp/](url)

`#!/bin/bash

#Get and save all homekit accessories to file (jq for better formatting)
curl -X PUT http://IP:PORT/accessories --header "Content-Type:Application/json" --header "authorization: \"031-45-154\"" -s | jq '.' > loghomekitaccessories.js#`

**Comment by cpoetter** [comment]
Thanks @Samfox2 ! I found out that the following command also shows the aid and iit

`DEBUG=* homebridge -D -I`

**Comment by zCrin** [comment]
Help I can't get accessory list with curl.... It says error 401 unauthorised. Please help !

**Comment by PierfrancescoElia** [comment]
Hello everybody, the command `curl -X PUT http://127.0.0.1:51826/characteristics --header ‘Content-Type:Application/json’ --header ‘authorization: 012-34-567’ --data ‘{ \’characteristics\": [{ \’aid\’: 2, \’iid\’: 9, \’value\’: 0}] }’` work well for me, but I have a simple question.. :)

I want to invert the value, if now is 1 => 0 and if now is 0=> 1.

How can I do that? Thanks

**Comment by zCrin** [comment]
You will alphabet to write a bash script that get the value and inverse it

**Comment by PierfrancescoElia** [comment]
@zCrin that would be great, so what is the command that returns the actual value?

And where I can find documentation for this PUT requests?

**Comment by PierfrancescoElia** [comment]
Found!

`curl -X GET http://127.0.0.1:51826/characteristics?id=2.9 --header "Content-Type:Application/json" --header "authorization: 012-34-567"`

Where 2.9 means:

- 2: AID

- 9: IID

-

**Comment by zCrin** [comment]
Interesting I didn't know this URL. How do you know which URL to call ?

**Comment by PierfrancescoElia** [comment]
I found that link by chance, as there isn't any documentation about that.
If you look [here](https://github.com/KhaosT/HAP-NodeJS/blob/3733e7d7274a656c71fe02fd3df4af170dd33c3b/lib/HAPServer.js#L650), you can read that if you try a GET request like /characteristic?id=aid1.iid1,aid2.iid2 you can get a JSON where you can find the actual state [0,1] (value field)..

So in my script I have a GET request to Homebridge, then analyze the response (value filed) and turn on or off the lamp! :)

**Comment by zCrin** [comment]
I've got a new issue; I have a temperature device but when I run 
`curl -X PUT ". $this->host . ":". $this->port ."/accessories --header \"Content-Type:application/json\" --header 'authorization: \"". $this->pin ."\"' -s"`
It retun the same value and I need to go in the iOS app to refresh the value

**Comment by agrublev** [comment]
For people having issues with suggested commands, here is how i got it to work:

I ran homebridge using:
`DEBUG=* homebridge -D -I
`

Then i clicked in my Home app on my iPhone to turn a light on or off 

In the console it showed me:
`EventedHTTPServer [::ffff:10.0.1.29] Sending HTTP event '19.8' with data: {"characteristics":[{"aid":19,"iid":8,"value":true}]} +0ms`

Which is the right iid to use.

Then i used
`curl -X PUT http://10.0.1.52:51852/characteristics --header "Content-Type:Application/json" --header "authorization: 031-45-152” --data "{\"characteristics\":[{\"aid\":2,\"iid\”:9,\”On\”:true}]}"`

Which turns it on! And to get a list of all lights
`curl -X GET http://10.0.1.52:51852/characteristics --header "Content-Type:Application/json" --header "authorization: 031-45-152” -s`

**Comment by qudarr** [comment]
For anybody wondering, why the curl statement above doesn't work for you, please check:

- [ ] Don't blindly copy & paste the above curl statement, as the apostrophes will not work correctly. You will need to re-type them. In my case: using chrome on windows as well as safari on mac.
- [ ] Do you have the same homebridge port? Mine was 51826.
- [ ] Check your authorization code
- [ ] I didn't need the "-s" parameter at the end of the curl GET statement

To get all of the available AID and IID values, execute following script:

```
#!/bin/bash
for ((i = 0; i < 100; i++))
do
  for ((j = 0; j < 100; j++))
  do
    echo "$i $j"
    curl --silent -X GET http://127.0.0.1:51826/characteristics?id=$i.$j --header "Content-Type:Application/json" --header 'authorization: XXX-YY-ZZZ'
  done
done
```

**Comment by NorthernMan54** [comment]
Another way to get all the devices, is just use the GET /accessories command in your curl instead, it will return all devices including a lot more details.

**Comment by qudarr** [comment]
Thanks, this is much better.

**Comment by felipesalomao** [comment]
Hi, i want make script that change device status to opposite when detect especific infrared signal (Hex code), i already am able to get device status, for example: TV
`curl -X GET http://localhost:51820/characteristics?id=4.10 --header "Content-Type:Application/json" --header "authorization: 031-45-155"`
I have broadlink rmpro that can send and receive ir commands, it have plugin for homebridege and works like charm: https://github.com/lprhodes/homebridge-broadlink-rm
To active it learn IR commands it i use: 
`curl -X PUT http://localhost:51820/characteristics --header "Content-Type:Application/json" --header "authorization:031-45-155" --data "{\"characteristics\":[{\"aid\":2,\"iid\":10,\"value\":true}]}"`

The problem is that it stop learn few seconds latter.. Also i would new get this hex code (i can see it on homebridge debug), another problem is that if it read same remote button, it aways is different hex code, its similar, but hard be exactly same.. 
Also could use if needed GPIO ir receiver.. 

----------------------

I think will be hard somebody here with experience in rm pro devices (their git is stoped, no updates long time, so is hard discuss with guys in issues list)

But at least could help with that:

`curl -X GET http://localhost:51820/characteristics?id=4.10 --header "Content-Type:Application/json" --header "authorization: 031-45-155"`

How make script to read it result (example bellow):

{"characteristics":[{"aid":4,"iid":10,"value":false,"status":0}]}
it have value  = false, i need to change that value to true and send command: 

`curl -X PUT http://localhost:51820/characteristics --header "Content-Type:Application/json" --header "authorization:031-45-155" --data "{\"characteristics\":[{\"aid\":4,\"iid\":10,\"value\":true}]}"`

**Comment by i3laze** [comment]
To summ up:
1. Make sure you run in Insecure Mode.
2. Output of _/accessories_ is rough to understand:
_curl -X GET --header "Content-Type:Application/json" --header "authorization: 031-45-154" http://YOUR_IP_HERE:51826/accessories_
So better use this one for discovery:
`for ((i = 0; i < 10; i++))
do
  for ((j = 0; j < 30; j++))
  do
    curl -s -X GET http://YOUR_IP_HERE::51826/characteristics?id=$i.$j --header "Content-Type:Application/json" --header 'authorization: 031-45-154' 2>&1 | grep value
  done
  printf "\n"
done`

**Comment by i3laze** [comment]
Seems so) mine was aid=2, with value contained instring was iid=10.
Run the discovery script to figure out own values.

**Comment by ebaauw** [comment]
They're like index numbers for the accessory (aid) and service/characteristic (iid), within the accessory.  I suppose to avoid having to pass lengthy UUIDs with every HAP call.  They're mapped to the UUIDs, see `~/.homebridge/persist/IdentifierCache.`_username_`.json`.

**Comment by christiansassano** [comment]
> Run the discovery script to figure out own values.

Thank you very much, and sorry for this message, but what discovery script can I run to figure out my values?
Sorry sorry

**Comment by christiansassano** [comment]
> They're mapped to the UUIDs, see ~/.homebridge/persist/IdentifierCache.username.json.

Thanks, I find the file but, into the file, there is so much number :-(

**Comment by ebaauw** [comment]
It's an object with key/values pairs, where
- the key is the accessory UUID and the value the aid;
- the key is the accessory UUID followed by `|`, followed by the service (type) UUID and `|` optionally followed by the service subtype and `|`, followed by the characteristic (type) UUID.

See https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js for the UUIDs of the standard services and characteristics.  
The accessory UUID is created by homebridge when the accessory is created: https://github.com/nfarina/homebridge/blob/d5691fd19e8b9395b3120500217178c4728bfade/lib/server.js#L404-L406

**Comment by christiansassano** [comment]
Thanks, I suppose to find my data
`.....ca|nextIID":12,`

`.....ca|nextIID":12,`

but onece aid

`....."|nextAID":19,`

Is correct?

**Comment by christiansassano** [comment]
I update my string:
curl -X PUT http://IP:PORT/characteristics --header "Content-Type:Application/json" --header "authorization:XXX-XX-XXX" --data "{\"characteristics\":[{\"aid\":19,\"iid\":12,\"value\":1}]}"
But: {"status":-70401}pi@raspberrypi:

**Comment by ebaauw** [comment]
No, these are just references to the next aid to be assigned and to the next iid to be assigned for the accessory.  You need to look for your accessory UUID (I wouldn't know how to retrieve that) and the service and characteristic UUIDs.

If you run `DEBUG=* homebridge -D`, you'll see debug messages of the HAP protocol including the aid and iid values.

For example, here's one of my Hue dimmer switches, as exposed by homebridge-hue:
```
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74": 187,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|nextIID": 37,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|0000003E-0000-1000-8000-0026BB765291|00000014-0000-1000-8000-0026BB765291": 2,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|0000003E-0000-1000-8000-0026BB765291|00000020-0000-1000-8000-0026BB765291": 3,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|0000003E-0000-1000-8000-0026BB765291|00000021-0000-1000-8000-0026BB765291": 4,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|0000003E-0000-1000-8000-0026BB765291|00000023-0000-1000-8000-0026BB765291": 5,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|0000003E-0000-1000-8000-0026BB765291|00000030-0000-1000-8000-0026BB765291": 6,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|0000003E-0000-1000-8000-0026BB765291|00000052-0000-1000-8000-0026BB765291": 7,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291": 8,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291|00000023-0000-1000-8000-0026BB765291": 9,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291|000000CD-0000-1000-8000-0026BB765291": 10,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291|00000023-0000-1000-8000-656261617577": 11,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291|00000077-0000-1000-8000-0026BB765291": 12,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291|00000075-0000-1000-8000-0026BB765291": 13,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291|00000022-0000-1000-8000-656261617577": 14,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|000000CC-0000-1000-8000-0026BB765291|00000021-0000-1000-8000-656261617577": 15,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|On": 16,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|On|00000023-0000-1000-8000-0026BB765291": 17,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|On|00000073-0000-1000-8000-0026BB765291": 18,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|On|000000CB-0000-1000-8000-0026BB765291": 19,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Up": 20,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Up|00000023-0000-1000-8000-0026BB765291": 21,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Up|00000073-0000-1000-8000-0026BB765291": 22,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Up|000000CB-0000-1000-8000-0026BB765291": 23,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Down": 24,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Down|00000023-0000-1000-8000-0026BB765291": 25,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Down|00000073-0000-1000-8000-0026BB765291": 26,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Dim Down|000000CB-0000-1000-8000-0026BB765291": 27,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Off": 28,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Off|00000023-0000-1000-8000-0026BB765291": 29,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Off|00000073-0000-1000-8000-0026BB765291": 30,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000089-0000-1000-8000-0026BB765291|Off|000000CB-0000-1000-8000-0026BB765291": 31,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000096-0000-1000-8000-0026BB765291": 32,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000096-0000-1000-8000-0026BB765291|00000023-0000-1000-8000-0026BB765291": 33,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000096-0000-1000-8000-0026BB765291|00000068-0000-1000-8000-0026BB765291": 34,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000096-0000-1000-8000-0026BB765291|0000008F-0000-1000-8000-0026BB765291": 35,
    "3979a1a8-fcb5-4abe-abea-d07911ecaa74|00000096-0000-1000-8000-0026BB765291|00000079-0000-1000-8000-0026BB765291": 36,
```
Parsing this:
- The aid is 187;
- The next iid to be assigned will be 37;
- I think iid 1 is reserved for the _Accessory Information_ service, hence not present in the file;
- iids 2-8 are the characteristics of the _Accessory Information_ service (03E): _Identify_ (014), _Manufacturer_ (020), _Model_ (021), _Name_ (023), _Serial Number_ (030), and _Firmware Revision_ (052);
- iid 8 is for the _Service Label_ service (0CC);
- iids 9, 10, 12, 13 are for standard characteristics of the _Service Label_ service: _Name_ (023), _Service Label Namespace_ (0CD), _Status Fault_ (077), _Status Active_ (075);
- iids 11, 14 and 15 are custom characteristics for the _Service Label_ service (see: https://github.com/ebaauw/homebridge-hue-utils/blob/master/lib/MyHomeKitTypes.js): _Last Updated (023), _Enabled_ (022), _Resource_ (021);
- iid 16 is the fist _Stateless Programmable Switch_ service (089) for the first button (subtype "On");
- iids 17-19 are the characteristics for the fist _Stateless Programmable Switch_ service: _Name_ (023), _Programmable Switch Event_ (073), and _Service Label Index_ (0CB);
- This repeats for the other three buttons;
- iid 32 is the _Battery_ (096) service;
- iids 33-36 are the characteristics fo the _Battery_ service: _Name_ (023), _Battery Level_ (068), _Charging State_ (08F), and _Status Low Battery_ (079).

**Comment by ebaauw** [comment]
> I update my string:
curl -X PUT http://IP:PORT/characteristics --header "Content-Type:Application/json" --header "authorization:XXX-XX-XXX" --data "{"characteristics":[{"aid":19,"iid":12,"value":1}]}"
But: {"status":-70401}pi@raspberrypi:

Did you run `homebridge -I`?

**Comment by christiansassano** [comment]
Yes, but there is an error:

`Or enter this code with your HomeKit app on your iOS device to pair with Homebridge:
                       
    ┌────────────┐     
    │ 012-55-234 │     
    └────────────┘     
                       
events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE :::51826
    at Object._errnoException (util.js:1022:11)
    at _exceptionWithHostPort (util.js:1044:20)
    at Server.setupListenHandle [as _listen2] (net.js:1351:14)
    at listenInCluster (net.js:1392:12)
    at Server.listen (net.js:1476:7)
    at EventedHTTPServer.listen (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/util/eventedhttp.js:60:19)
    at HAPServer.listen (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/HAPServer.js:158:20)
    at Bridge.Accessory.publish (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:607:16)
    at Server._publish (/usr/local/lib/node_modules/homebridge/lib/server.js:126:16)
    at Server.run (/usr/local/lib/node_modules/homebridge/lib/server.js:96:10)`

**Comment by christiansassano** [comment]
I suppose because I use this for a quick install of Homebridge server:

https://itunes.apple.com/it/app/homebridge-for-raspberrypi/id1123183713?mt=8

I think the structure of the directory is different :-(

**Comment by ebaauw** [comment]
> Error: listen EADDRINUSE :::51826

You're already/still running homebridge.  Make sure to stop any service before starting homebridge manually from the command line.

**Comment by i3laze** [comment]
The script by @qudarr, which I posted in step 2, provides a very clear aid/uid output.
https://github.com/nfarina/homebridge/issues/506#issuecomment-387980947

**Comment by i3laze** [comment]
IMHO, HomeBridge developers should publish some straightforward CLI script or _more_ simplified API for this Issue (listing devices and switching their state).
It's much easier to control the Tasmota device via http. Not using HomeBridge at all.
My whole toggle script is now:
`#!/bin/sh
IP=192.168.1.129

result=$( curl -s -X GET http://$IP/cm?cmnd=Power | jq -r '.POWER?' )
echo $result

case $result in
  ON)
    echo "Toggling off.."
    curl -s -X GET http://$IP/cm?cmnd=Power%20Off ;;
  OFF)
    echo "Toggling on.."
    curl -s -X GET http://$IP/cm?cmnd=Power%20On ;;
  *)
    echo "Invalid cURL response!"
    exit 1 ;;
esac
`

**Comment by i3laze** [comment]
Looks like there were some changes in Apple HAP protocol, that HomeBridge exposes with these cURL requests.  My values doesn't follow toggle state anymore.

Unfortunately, access to HAP protocol specification now unavailable/requires MFi enrollment.

**Comment by NorthernMan54** [comment]
@i3laze I’m a big user of insecure mode and have not noticed anything recently.  It is the backbone of my Alexa plugin, and has been working for over a year.  Can you elaborate further on your issue?

**Comment by i3laze** [comment]
My fault..  seems every device has own _aid_ -
A year ago I thought devices are placed under single _aid_ (between 1 and 3)..

```
14 9; {"characteristics":[{"aid":14,"iid":9,"value":"Kodi Playback","status":0}]}
14 10; {"characteristics":[{"aid":14,"iid":10,"value":true,"status":0}]}
```

Seems, parsing _curl -s -X GET http://localhost:51826/accessories_ is the only reliable solution.

**Comment by NorthernMan54** [comment]
If your playing with this interface, I have a couple of projects that use it heavily.  Take a look at my homebridge-Alexa plugin and Hap-NodeRed repo’s.
